js/okcoinusd.js

```diff
@@ -623,17 +623,29 @@ module.exports = class okcoinusd extends Exchange {
         return { 'url': url, 'method': method, 'body': body, 'headers': headers };
     }
 
-    async request (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
-        let response = await this.fetch2 (path, api, method, params, headers, body);
+    handleErrors (code, reason, url, method, headers, body) {
+        if (this.verbose)
+            console.log (this.id, method, url, code, reason, body ? ("\nResponse:\n" + body) : '');
+        let response = JSON.parse (body);
+        if ('error_code' in response) {
+            if (this.errorCodes === undefined) {
+                this.errorCodes = {
+                    '1009': OrderNotFound,
+                    '1003': InvalidOrder,      // no order type (was left by previous author)
+                    '1027': InvalidOrder,      // createLimitBuyOrder(symbol, 0, 0): Incorrect parameter may exceeded limits
+                    '1002': InsufficientFunds, // The transaction amount exceed the balance
+                    '10000': ExchangeError,    // createLimitBuyOrder(symbol, undefined, undefined)
+                    '10008': ExchangeError,    // Illegal URL parameter
+                };
+            }
+            let Exception = this.errorCodes[response['error_code']];
+            if (Exception)
+                throw new Exception (this.id + ' ' + this.json (response));
+            else
+                throw new ExchangeError (this.id + ' ' + this.json (response));
```

Comment:
Also, I'm not sure if this particular construct will work well in all languages. Most likely will have to rewrite this for if-else nevertheless. Currently the transpiler can only guess proper throw syntax within one line, no code lookups are done, so it's dump thin converter (for now).

Reply:
Oh, my God, it's going to be UGLY. There are just six checks at the moment but as the library is gaining usage we'll be discovering more of them. I won't be surprised if we end up with dozens of cases and those `if`s will span pages of code.

Is it possible to do a helper for all three languages? Kind of "ThrowableMap".

@kornrunner, I see you do a lot of contribution on PHP side. I'm not an expert on that language (actually, I'm not an expert in JS too :), but how would you address this issue?

Basically, there are exchanges that report errors by codes and those would be ideal to handle via a `switch` statement. Considering, we can't use `switch`, how would you address such a task in PHP?

https://github.com/mkutny/ccxt/blob/a43b3fdc22696fb40120174e59e9019098460a4b/js/okcoinusd.js#L630-L645

Reply:
I can tell right away, this is ok in PHP (variables = exception classes). I'm not sure on Python, in that particular case it will definitely throw, because Exception is a reserved keyword for the base exception class there.